Fix tx resend

LyzeOfKiel opened the pull request:
Closes #35 
Added assertion for permit deadline when resending a transaction.
https://github.com/zkBob/zeropool-relayer/blob/68e93a906abc45469a7be9f541f04d6c5bb3f728/zp-relayer/workers/sentTxWorker.ts#L137-L140

Closes #49 
Added handling of some client errors when resending a transaction.
https://github.com/zkBob/zeropool-relayer/blob/68e93a906abc45469a7be9f541f04d6c5bb3f728/zp-relayer/workers/sentTxWorker.ts#L177
https://github.com/zkBob/zeropool-relayer/blob/68e93a906abc45469a7be9f541f04d6c5bb3f728/zp-relayer/workers/sentTxWorker.ts#L180

Closes #51 
Gas price is updated if corresponding error was thrown during transaction re-send. Related to #49 

Closes #52 
See #77 

Closes #74 
Implemented proper transaction re-send mechanism. Re-send errors related to gasPrice/nonce are handled. Other errors are considered as unrecoverable, and in this case optimistic state rollback is performed.
https://github.com/zkBob/zeropool-relayer/blob/68e93a906abc45469a7be9f541f04d6c5bb3f728/zp-relayer/workers/sentTxWorker.ts#L183-L187

Closes #77 
Added `MAX_FEE_PER_GAS_LIMIT` env variable for limiting max value of `maxFeePerGas`.
Gas parameters are calculated as follows:
`new_maxFeePerGas = min(MAX_FEE_PER_GAS_LIMIT, old_maxFeePerGas)`
`new_maxPriorityFeePerGas = max(0, old_maxPriorityFeePerGas - (old_maxFeePerGas - MAX_FEE_PER_GAS_LIMIT))` 
This behaviour is described in tests.

k1rill-fedoseev commented on zp-relayer/validateTx.ts:
```diff
@@ -233,7 +231,13 @@ export async function validateTx({ txType, rawMemo, txProof, depositSignature }:
   }
+  if (txType === TxType.PERMITTABLE_DEPOSIT) {
+    const deadline = (txData as PermittableDepositTxData).deadline
+    await checkAssertion(() => checkDeadline(deadline))
```
I think we should also configure some buffer window for checking the deadline. e.g. we should reject transactions, whose deadline will expire in 5 seconds, since it won't be mined in such short time in any case.

I would suggest the following:
* During the initial transaction sending, we should reject transaction with deadline in less than 10 minutes
* During transaction resend, we should reject transaction with deadline in less than 10 seconds

LyzeOfKiel replied:
Agree. Maybe it is better to create a special env (e.g. `PERMIT_DEADLINE_RESEND_WINDOW`) for this deadline window for resend case? And calculate initial sending window as `PERMIT_DEADLINE_RESEND_WINDOW + 10 min`.

k1rill-fedoseev replied:
Current threshold in library is set to 5 minutes, so let's stick to that value for now, for the initial tx sending check - https://github.com/zkBob/zkbob-client-js/pull/34

LyzeOfKiel commented on zp-relayer/services/gas-price/GasPrice.ts:
```diff
@@ -26,6 +29,43 @@ const polygonGasPriceKeyMapping: Record<GasPriceKey, PolygonGSV2GasPriceKey> = {
+    return fees
+  } else {
+    const maxFeePerGas = maxFeeLimit.toString(10)
+    const maxPriorityFeePerGas = BN.max(toBN(fees.maxPriorityFeePerGas).sub(diff), toBN(0)).toString(10)
```
Agree

LyzeOfKiel commented:
New changes:
* Added two variables for threshold configuration:
`PERMIT_DEADLINE_THRESHOLD_INITIAL` (default: 300)
`PERMIT_DEADLINE_THRESHOLD_RESEND` (default: 10)
* Added test `checkDeadline` method with thresholds
* Refactoring

k1rill-fedoseev commented on zp-relayer/workers/sentTxWorker.ts:
```diff
@@ -99,55 +124,81 @@ export async function createSentTxWorker<T extends EstimationType>(gasPrice: Gas
+      // Resend with updated gas price
+      if (txType === TxType.PERMITTABLE_DEPOSIT) {
+        const deadline = (txData as PermittableDepositTxData).deadline
+        await checkAssertion(() => checkDeadline(deadline, config.permitDeadlineThresholdResend))
```
Can you clarify, what will happen with the this particular job if the deadline will be expired at this point? What will happen with the jobs that are positioned later in the optimistic state?

LyzeOfKiel replied:
This part with deadline check should be moved inside try-catch block below (just before transaction submission), so that this error will be caught and state rollback is performed. Will fix now.

LyzeOfKiel commented:
Changes:
* Moved deadline validation in try-catch block. If validation will fail, all succeeding optimistic transactions will be erased and optimistic state will be cleared.
* Changed some `BN` types to `string`, as they potentially can cause errors in queue-worker communication.

k1rill-fedoseev reviewed (approved):
lgtm

I hope this cases will be well-covered by the e2e auto tests in the future...
